File: ios/Classes/InputSignalCollector.swift
import Foundation
import UIKit

/// Collects input signals like keystroke timing.
/// Privacy: NO text content is collected, only timing metrics.
class InputSignalCollector: NSObject, UITextFieldDelegate, UITextViewDelegate {

    private var config: BehaviorConfig
    private var eventHandler: ((BehaviorEvent) -> Void)?
    private var keystrokeTimestamps: [Double] = []
    private var lastKeystrokeTime: Double = 0
    private var currentBurstLength: Int = 0
    private let maxBurstGap: Double = 2.0 // 2 seconds

    private var monitoredTextFields: [UITextField] = []
    private var monitoredTextViews: [UITextView] = []

    init(config: BehaviorConfig) {
        self.config = config
        super.init()
        setupNotifications()
    }

    func setEventHandler(_ handler: @escaping (BehaviorEvent) -> Void) {
        self.eventHandler = handler
    }

    func attachToView(_ view: UIView) {
        guard config.enableInputSignals else { return }

        if let textField = view as? UITextField {
            textField.delegate = self
            monitoredTextFields.append(textField)
            NotificationCenter.default.addObserver(
                self,
                selector: #selector(textFieldDidChange(_:)),
                name: UITextField.textDidChangeNotification,
                object: textField
            )
        } else if let textView = view as? UITextView {
            textView.delegate = self
            monitoredTextViews.append(textView)
            NotificationCenter.default.addObserver(
                self,
                selector: #selector(textViewNotificationDidChange(_:)),
                name: UITextView.textDidChangeNotification,
                object: textView
            )
        }

        // Recursively attach to subviews
        for subview in view.subviews {
            attachToView(subview)
        }
    }

    func updateConfig(_ newConfig: BehaviorConfig) {
        config = newConfig
    }

    func dispose() {
        NotificationCenter.default.removeObserver(self)
        keystrokeTimestamps.removeAll()
        monitoredTextFields.removeAll()
        monitoredTextViews.removeAll()
    }

    private func setupNotifications() {
        // Monitor all text input changes globally
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(textDidChange(_:)),
            name: UITextField.textDidChangeNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(textDidChange(_:)),
            name: UITextView.textDidChangeNotification,
            object: nil
        )
    }

    @objc private func textFieldDidChange(_ notification: Notification) {
        onKeystroke()
    }

    @objc private func textViewNotificationDidChange(_ notification: Notification) {
        onKeystroke()
    }

    @objc private func textDidChange(_ notification: Notification) {
        onKeystroke()
    }

    private func onKeystroke() {
        let now = Date().timeIntervalSince1970 * 1000 // milliseconds
        keystrokeTimestamps.append(now)

        // Keep only last 100 keystrokes
        if keystrokeTimestamps.count > 100 {
            keystrokeTimestamps.removeFirst()
        }

        // Check if part of current burst
        if lastKeystrokeTime > 0 && (now - lastKeystrokeTime) < maxBurstGap * 1000 {
            currentBurstLength += 1
        } else {
            // New burst
            if currentBurstLength > 0 {
                emitTypingBurst(burstLength: currentBurstLength)
            }
            currentBurstLength = 1
        }

        // Calculate inter-key latency and emit typing cadence
        // For first keystroke, use a default latency of 0 (will be counted but with 0 latency)
        if lastKeystrokeTime > 0 {
            let latency = now - lastKeystrokeTime
            emitTypingCadence(interKeyLatency: latency)
        } else {
            // First keystroke: emit with 0 latency so it's counted
            emitTypingCadence(interKeyLatency: 0)
        }

        lastKeystrokeTime = now
    }

    private func emitTypingCadence(interKeyLatency: Double) {
        // In new model, keystrokes are tracked as tap events
        // Estimate tap duration (typically 50-150ms for keyboard taps)
        let estimatedTapDuration = Int(max(50, min(150, interKeyLatency)))

        eventHandler?(BehaviorEvent(
            sessionId: "current",
            eventType: "tap",
            metrics: [
                "tap_duration_ms": estimatedTapDuration,
                "long_press": false
            ]
        ))
    }

    private func emitTypingBurst(burstLength: Int) {
        // Bursts are now tracked as individual tap events
        // No need to emit separate burst events
    }

    // MARK: - UITextFieldDelegate

    func textFieldDidBeginEditing(_ textField: UITextField) {
        // Text field focused
    }

    func textFieldDidEndEditing(_ textField: UITextField) {
        // Text field unfocused
    }

    // MARK: - UITextViewDelegate

    func textViewDidBeginEditing(_ textView: UITextView) {
        // Text view focused
    }

    func textViewDidEndEditing(_ textView: UITextView) {
        // Text view unfocused
    }
}

File: android/src/main/java/ai/synheart/behavior/InputSignalCollector.kt
package ai.synheart.behavior

import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import java.util.LinkedList

/**
 * Collects input signals like keystroke timing.
 * Privacy: NO text content is collected, only timing metrics.
 */
class InputSignalCollector(private var config: BehaviorConfig) {

    private var eventHandler: ((BehaviorEvent) -> Unit)? = null
    private val keystrokeTimestamps = LinkedList<Long>()
    private var lastKeystrokeTime: Long = 0
    private var currentBurstLength = 0
    private val maxBurstGap = 2000L // 2 seconds between keystrokes to be in same burst

    private val textWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            if (count > 0) { // Character added
                onKeystroke()
            }
        }

        override fun afterTextChanged(s: Editable?) {}
    }

    fun setEventHandler(handler: (BehaviorEvent) -> Unit) {
        this.eventHandler = handler
    }

    fun attachToView(view: View) {
        if (!config.enableInputSignals) return

        when (view) {
            is EditText -> {
                view.addTextChangedListener(textWatcher)
            }
            is ViewGroup -> {
                // Recursively attach to all EditText children
                for (i in 0 until view.childCount) {
                    attachToView(view.getChildAt(i))
                }
            }
        }
    }

    fun updateConfig(newConfig: BehaviorConfig) {
        config = newConfig
    }

    private fun onKeystroke() {
        val now = System.currentTimeMillis()
        keystrokeTimestamps.add(now)

        // Keep only last 100 keystrokes
        while (keystrokeTimestamps.size > 100) {
            keystrokeTimestamps.removeFirst()
        }

        // Check if part of current burst
        if (lastKeystrokeTime > 0 && (now - lastKeystrokeTime) < maxBurstGap) {
            currentBurstLength++
        } else {
            // New burst
            if (currentBurstLength > 0) {
                emitTypingBurst(currentBurstLength)
            }
            currentBurstLength = 1
        }

        // Calculate inter-key latency
        if (lastKeystrokeTime > 0) {
            val latency = now - lastKeystrokeTime
            emitTypingCadence(latency)
        }

        lastKeystrokeTime = now
    }

    private fun emitTypingCadence(interKeyLatency: Long) {
        // Calculate rolling cadence (keys per second)
        val recentKeys = keystrokeTimestamps.filter { it > System.currentTimeMillis() - 5000 }
        val cadence = if (recentKeys.size > 1) {
            val timeSpan = recentKeys.last() - recentKeys.first()
            if (timeSpan > 0) (recentKeys.size - 1) * 1000.0 / timeSpan else 0.0
        } else {
            0.0
        }

        eventHandler?.invoke(
            BehaviorEvent(
                sessionId = "current", // Will be set by SDK
                timestamp = System.currentTimeMillis(),
                type = "typingCadence",
                payload = mapOf(
                    "cadence" to cadence,
                    "inter_key_latency" to interKeyLatency,
                    "keys_in_window" to recentKeys.size
                )
            )
        )
    }

    private fun emitTypingBurst(burstLength: Int) {
        if (burstLength < 3) return // Only emit significant bursts

        val recentLatencies = mutableListOf<Long>()
        for (i in 1 until keystrokeTimestamps.size.coerceAtMost(burstLength)) {
            recentLatencies.add(keystrokeTimestamps[i] - keystrokeTimestamps[i - 1])
        }

        val avgLatency = if (recentLatencies.isNotEmpty()) {
            recentLatencies.average()
        } else 0.0

        val variance = if (recentLatencies.size > 1) {
            val mean = recentLatencies.average()
            recentLatencies.map { (it - mean) * (it - mean) }.average()
        } else 0.0

        eventHandler?.invoke(
            BehaviorEvent(
                sessionId = "current",
                timestamp = System.currentTimeMillis(),
                type = "typingBurst",
                payload = mapOf(
                    "burst_length" to burstLength,
                    "inter_key_latency" to avgLatency,
                    "variance" to variance
                )
            )
        )
    }

    fun dispose() {
        keystrokeTimestamps.clear()
    }
}
